### PR TITLE
event crawler added

### PR DESCRIPTION
**Pull request by RAIL-HU**
Failed to test locally, BeautifulSoup 4 added in requirements.txt but failed to be imported; how to handle the edge case when the time is inside an event is unconfirmed (display only the current event or both current and next)

Potential risks of the crawler: the detection of event table is based on finding <td> with specific style sheets in the HTML document, if they are changed in the future, the only apparent sign is the crawler fails to work. Manual update is required to make it continue to work properly

### CODE REVIEW

**Comment by EricPedley** [comment]
Just pushed two commits to make it work. Now gonna start reviewing for code quality and possibly requesting changes or clarification.

File: api/location_management.py [hunk 1]
```diff
@@ -41,3 +43,48 @@ def get_schedule_data(location, date):
...
+
+    soup = bs(html, 'html.parser')
+    table = soup.findAll('td', {'style': ['padding: 0.75pt; border-width: 0.75pt; border-right-style: dashed; border-right-color: #999999; border-bottom-style: dashed; border-bottom-color: #999999; text-align: left;',\
+                                        'padding: 0.75pt; border-width: 0.75pt; border-right-style: dashed; border-right-color: #999999; border-bottom-style: dashed; border-bottom-color: #999999; text-align: center;',\
+                                        'border-right: 0.75pt dashed #999999; border-bottom: 0.75pt dashed #999999; text-align: left;',\
+                                        'padding: 0.75pt; border-right: 0.75pt dashed #999999; text-align: left;',\
+                                        'padding: 0.75pt; border-bottom: 0.75pt dashed #999999; text-align: left;']})
+    entries_list = list()
+
+    for i in table:
```
Comment (EricPedley) [inline api/location_management.py:67]: This would be a good place to use the built-in `map` function instead of a loop.

File: api/location_management.py [hunk 2]
```diff
@@ -41,3 +43,48 @@ def get_schedule_data(location, date):
...
+        item_text = i.getText().strip('\n')
+        entries_list.append(item_text)
+
+    headers = ('date', 'name', 'service_start', 'service_end')
```
Comment (EricPedley) [inline api/location_management.py:71]: Putting these string literals into a tuple isn't very useful and decreases readability. I would replace the references to this tuple with the string literals themselves.

File: api/location_management.py [hunk 3]
```diff
@@ -41,3 +43,48 @@ def get_schedule_data(location, date):
...
+    rows_list = entries_list[5:]
+
+    row = {}
+    curr_time = get_irvine_time()
+    for i in range(0, int(len(rows_list) / 4)):
```
Comment (EricPedley) [inline api/location_management.py:76]: What does this loop do? I would think we wouldn't need a loop here as currently the function only returns one row.

**Review by EricPedley — CHANGES_REQUESTED** [review]

File: api/location_management.py (hunk 3 above)
Comment (RAIL-HU) [reply api/location_management.py:76]: It goes through all the event data to filter out the one closest to current time

File: api/location_management.py (hunk 1 above)
Comment (RAIL-HU) [reply api/location_management.py]: map function is used in the latest commit

File: api/location_management.py (hunk 2 above)
Comment (RAIL-HU) [reply api/location_management.py]: the tuple is removed and changed key of dict to string literals in the latest commit

**Review by EricPedley — APPROVED** [review]
Looks good. We might eventually end up refactoring the loop but it's good enough for now.

**Comment by EricPedley** [comment]
Ended up modifying the method a little.
Check out this commit (https://github.com/EricPedley/zotmeal-backend/commit/6d746c45eadf2b42ef16f10ea7042e519edd866e), the current head of main, and take note of the different design decisions and/or tell me if you have any concerns.
